File: android/src/main/java/com/polidea/flutter_ble_lib/FlutterBleLibPlugin.java
package com.polidea.flutter_ble_lib;

import android.content.Context;
import android.util.Log;

import com.polidea.flutter_ble_lib.constant.ArgumentKey;
import com.polidea.flutter_ble_lib.constant.ChannelName;
import com.polidea.flutter_ble_lib.constant.MethodName;
import com.polidea.flutter_ble_lib.delegate.BluetoothStateDelegate;
import com.polidea.flutter_ble_lib.delegate.CallDelegate;
import com.polidea.flutter_ble_lib.delegate.CharacteristicsDelegate;
import com.polidea.flutter_ble_lib.delegate.DescriptorsDelegate;
import com.polidea.flutter_ble_lib.delegate.DeviceConnectionDelegate;
import com.polidea.flutter_ble_lib.delegate.DevicesDelegate;
import com.polidea.flutter_ble_lib.delegate.LogLevelDelegate;
import com.polidea.flutter_ble_lib.delegate.DiscoveryDelegate;
import com.polidea.flutter_ble_lib.delegate.MtuDelegate;
import com.polidea.flutter_ble_lib.delegate.RssiDelegate;
import com.polidea.flutter_ble_lib.event.AdapterStateStreamHandler;
import com.polidea.flutter_ble_lib.event.ConnectionStateStreamHandler;
import com.polidea.flutter_ble_lib.event.RestoreStateStreamHandler;
import com.polidea.flutter_ble_lib.event.ScanningStreamHandler;
import com.polidea.multiplatformbleadapter.BleAdapter;
import com.polidea.multiplatformbleadapter.BleAdapterFactory;
import com.polidea.multiplatformbleadapter.OnErrorCallback;
import com.polidea.multiplatformbleadapter.OnEventCallback;
import com.polidea.multiplatformbleadapter.ScanResult;
import com.polidea.multiplatformbleadapter.errors.BleError;

import java.util.LinkedList;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.EventChannel;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

public class FlutterBleLibPlugin implements FlutterPlugin, MethodCallHandler {

    static final String TAG = FlutterBleLibPlugin.class.getName();

    private AdapterStateStreamHandler adapterStateStreamHandler = new AdapterStateStreamHandler();
    private RestoreStateStreamHandler restoreStateStreamHandler = new RestoreStateStreamHandler();
    private ScanningStreamHandler scanningStreamHandler = new ScanningStreamHandler();
//    private ConnectionStateStreamHandler connectionStateStreamHandler = new ConnectionStateStreamHandler();
//    private CharacteristicsMonitorStreamHandler characteristicsMonitorStreamHandler = new CharacteristicsMonitorStreamHandler();

    private List<CallDelegate> delegates = new LinkedList<>();

    @Nullable
    private BleAdapter bleAdapter;
    @Nullable
    private Context context;
    @Nullable
    private MethodChannel methodChannel;
    @Nullable
    private BinaryMessenger binaryMessenger;
    @Nullable
    static private FlutterBleLibPlugin plugin;

    @SuppressWarnings("deprecation")
    public static void registerWith(Registrar registrar) {
        FlutterBleLibPlugin p = new FlutterBleLibPlugin();
        p.context = registrar.context();
        plugin = p;
        plugin.startListening(registrar.messenger());
    }

    @Override
    public void onAttachedToEngine(@NonNull FlutterPluginBinding binding) {
        context = binding.getApplicationContext();
        startListening(
            binding.getBinaryMessenger()
        );
    }

    @Override
    public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
        stopListening();
    }

    private void startListening(BinaryMessenger messenger) {
        binaryMessenger = messenger;
        methodChannel = new MethodChannel(messenger, ChannelName.FLUTTER_BLE_LIB);

        final EventChannel bluetoothStateChannel = new EventChannel(messenger, ChannelName.ADAPTER_STATE_CHANGES);
        final EventChannel restoreStateChannel = new EventChannel(messenger, ChannelName.STATE_RESTORE_EVENTS);
        final EventChannel scanningChannel = new EventChannel(messenger, ChannelName.SCANNING_EVENTS);
//        final EventChannel connectionStateChannel = new EventChannel(messenger, ChannelName.CONNECTION_STATE_CHANGE_EVENTS);
//        final EventChannel characteristicMonitorChannel = new EventChannel(messenger, ChannelName.MONITOR_CHARACTERISTIC);


        methodChannel.setMethodCallHandler(this);

        scanningChannel.setStreamHandler(scanningStreamHandler);
        bluetoothStateChannel.setStreamHandler(plugin.adapterStateStreamHandler);
        restoreStateChannel.setStreamHandler(plugin.restoreStateStreamHandler);
//        connectionStateChannel.setStreamHandler(plugin.connectionStateStreamHandler);
//        characteristicMonitorChannel.setStreamHandler(plugin.characteristicsMonitorStreamHandler);
    }

    private void stopListening() {
        methodChannel.setMethodCallHandler(null);
        methodChannel = null;
    }

    public FlutterBleLibPlugin() {
    }

    private void setupAdapter(Context context) {
        bleAdapter = BleAdapterFactory.getNewAdapter(context);
        delegates.add(new DeviceConnectionDelegate(bleAdapter, binaryMessenger));
        delegates.add(new LogLevelDelegate(bleAdapter));
        delegates.add(new DiscoveryDelegate(bleAdapter));
        delegates.add(new BluetoothStateDelegate(bleAdapter));
        delegates.add(new RssiDelegate(bleAdapter));
        delegates.add(new MtuDelegate(bleAdapter));
        delegates.add(new CharacteristicsDelegate(bleAdapter, binaryMessenger));
        delegates.add(new DevicesDelegate(bleAdapter));
        delegates.add(new DescriptorsDelegate(bleAdapter));
    }

    @Override
    public void onMethodCall(MethodCall call, Result result) {
        Log.d(TAG, "on native side observed method: " + call.method);
        for (CallDelegate delegate : delegates) {
            if (delegate.canHandle(call)) {
                delegate.onMethodCall(call, result);
                return;
            }
        }

        switch (call.method) {
            case MethodName.CREATE_CLIENT:
                createClient(call, result);
                break;
            case MethodName.DESTROY_CLIENT:
                destroyClient(result);
                break;
            case MethodName.START_DEVICE_SCAN:
                startDeviceScan(call, result);
                break;
            case MethodName.STOP_DEVICE_SCAN:
                stopDeviceScan(result);
                break;
            case MethodName.IS_CLIENT_CREATED:
                isClientCreated(result);
                break;
            default:
                result.notImplemented();
        }
    }

    private void isClientCreated(Result result) {
        result.success(bleAdapter != null);
    }

    private void createClient(MethodCall call, Result result) {
        if (bleAdapter != null) {
            Log.w(TAG, "Overwriting existing native client. Use BleManager#isClientCreated to check whether a client already exists.");
        }
        setupAdapter(context);
        bleAdapter.createClient(call.<String>argument(ArgumentKey.RESTORE_STATE_IDENTIFIER),
                new OnEventCallback<String>() {
                    @Override
                    public void onEvent(String adapterState) {
                        adapterStateStreamHandler.onNewAdapterState(adapterState);
                    }
                }, new OnEventCallback<Integer>() {
                    @Override
                    public void onEvent(Integer restoreStateIdentifier) {
                        restoreStateStreamHandler.onRestoreEvent(restoreStateIdentifier);
                    }
                });
        result.success(null);
    }

    private void destroyClient(Result result) {
        if (bleAdapter != null) {
            bleAdapter.destroyClient();
        }
        scanningStreamHandler.onComplete();
        bleAdapter = null;
        delegates.clear();
        result.success(null);
    }

    private void startDeviceScan(@NonNull MethodCall call, Result result) {
        if (bleAdapter == null) {
            result.success(null);
            return;
        }
        List<String> uuids = call.<List<String>>argument(ArgumentKey.UUIDS);
        bleAdapter.startDeviceScan(uuids.toArray(new String[uuids.size()]),
                call.<Integer>argument(ArgumentKey.SCAN_MODE),
                call.<Integer>argument(ArgumentKey.CALLBACK_TYPE),
                new OnEventCallback<ScanResult>() {
                    @Override
                    public void onEvent(ScanResult data) {
                        scanningStreamHandler.onScanResult(data);
                    }
                }, new OnErrorCallback() {
                    @Override
                    public void onError(BleError error) {
                        scanningStreamHandler.onError(error);
                    }
                });
        result.success(null);
    }

    private void stopDeviceScan(Result result) {
        if (bleAdapter != null) {
            bleAdapter.stopDeviceScan();
        }
        scanningStreamHandler.onComplete();
        result.success(null);
    }


}

File: android/src/main/kotlin/hu/privatdoktor/flutter_ble_lib/FlutterBleLibPlugin.kt
package hu.privatdoktor.flutter_ble_lib

import android.util.Log
import hu.privatdoktor.flutter_ble_lib.event.AdapterStateStreamHandler
import hu.privatdoktor.flutter_ble_lib.event.RestoreStateStreamHandler
import hu.privatdoktor.flutter_ble_lib.event.ScanningStreamHandler
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.FlutterPlugin.FlutterPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import kotlinx.coroutines.*
import java.util.*
import kotlin.coroutines.suspendCoroutine


private fun UUIDfrom(bluetoothUUIDStr: String) : UUID {
    val baseUUIDPrefix = "0000"
    val baseUUIDSuffix = "-0000-1000-8000-00805F9B34FB"

    val outUUUID =
        if (bluetoothUUIDStr.length == 4) {
            baseUUIDPrefix + bluetoothUUIDStr + baseUUIDSuffix
        } else if (bluetoothUUIDStr.length == 8) {
            bluetoothUUIDStr + baseUUIDSuffix
        } else {
            bluetoothUUIDStr
        }

    return UUID.fromString(outUUUID)
}

class FlutterBleLibPlugin : FlutterPlugin, MethodCallHandler {

    private var client: Client? = null
    private var methodChannel: MethodChannel? = null

    companion object {
        const val TAG = "FlutterBleLibPlugin"
    }

    override fun onAttachedToEngine(binding: FlutterPluginBinding) {
        client = Client(binding)

        val methodChannel = MethodChannel(binding.binaryMessenger, ChannelName.FLUTTER_BLE_LIB)
        methodChannel.setMethodCallHandler(this)
        this.methodChannel = methodChannel
    }

    override fun onDetachedFromEngine(binding: FlutterPluginBinding) {
        methodChannel?.setMethodCallHandler(null)
        methodChannel = null
        client = null
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        Log.d(TAG, "on native side observed method: " + call.method)
        val client = this.client
        if (client == null) {
            Log.d(TAG, "on native side onMethodCall:: client was null ")
            return
        }
        GlobalScope.launch(Dispatchers.Main.immediate) {
            try {
                when (call.method) {
                    MethodName.IS_CLIENT_CREATED -> {
                        val isCreated = client.isClientCreated()
                        result.success(isCreated)
                    }
                    MethodName.CREATE_CLIENT -> {
                        client.createClient()
                        result.success(null)
                    }
                    MethodName.DESTROY_CLIENT -> {
                        client.destroyClient()
                        result.success(null)
                    }
                    MethodName.START_DEVICE_SCAN -> {
                        val filteredUUIDs =
                            call.argument<List<String>>(ArgumentKey.UUIDS)!!.map {
                                UUIDfrom(bluetoothUUIDStr = it)
                            }

                        client.startDeviceScan(
                            scanMode = call.argument<Int>(ArgumentKey.SCAN_MODE)!!,
                            callbackType = call.argument<Int>(ArgumentKey.CALLBACK_TYPE)!!,
                            filteredUUIDs = filteredUUIDs
                        )
                        result.success(null)
                    }
                    MethodName.STOP_DEVICE_SCAN -> {
                        client.stopDeviceScan()
                        result.success(null)
                    }
                    MethodName.ENABLE_RADIO -> {
                        client.enableRadio()
                        result.success(null)

                    }
                    MethodName.DISABLE_RADIO -> {
                        client.disableRadio()
                        result.success(null)

                    }
                    MethodName.GET_STATE -> {
                        val stateStr = client.getState()
                        result.success(stateStr)
                    }
                    MethodName.GET_AUTHORIZATION -> {
                        result.success("allowedAlways")
                    }
                    MethodName.CONNECT_TO_DEVICE -> {
                        val timeoutMillis =
                            try {
                                val unwrappedValue = call.argument<Int>(ArgumentKey.TIMEOUT_MILLIS)
                                unwrappedValue?.toLong()
                            } catch (exception: ClassCastException) {
                                call.argument<Long>(ArgumentKey.TIMEOUT_MILLIS)
                            }
                        client.connectToDevice(
                            deviceIdentifier = call.argument<String>(ArgumentKey.DEVICE_IDENTIFIER)!!,
                            isAutoConnect = call.argument<Boolean>(ArgumentKey.IS_AUTO_CONNECT),
                            requestMtu = call.argument<Int>(ArgumentKey.REQUEST_MTU),
                            refreshGatt = call.argument<Boolean>(ArgumentKey.REFRESH_GATT)!!,
                            timeoutMillis = timeoutMillis
                        )
                        result.success(null)
                    }
                    MethodName.IS_DEVICE_CONNECTED -> {
                        val isConnected = client.isDeviceConnected(
                            deviceIdentifier = call.argument<String>(ArgumentKey.DEVICE_IDENTIFIER)!!
                        )
                        result.success(isConnected)
                    }
                    MethodName.OBSERVE_CONNECTION_STATE -> {
                        val streamHandlerName = client.observeConnectionState(
                            deviceIdentifier = call.argument<String>(ArgumentKey.DEVICE_IDENTIFIER)!!,
                            emitCurrentValue = call.argument<Boolean>(ArgumentKey.EMIT_CURRENT_VALUE)!!
                        )
                        result.success(streamHandlerName)
                    }
                    MethodName.CANCEL_CONNECTION -> {
                        client.cancelConnection(
                            deviceIdentifier = call.argument<String>(ArgumentKey.DEVICE_IDENTIFIER)!!
                        )
                        result.success(null)
                    }
                    MethodName.DISCOVER_SERVICES -> {
                        val serviceUuids =
                            call.argument<List<String>>(ArgumentKey.SERVICE_UUIDS)?.map {
                                UUIDfrom(bluetoothUUIDStr = it)
                            }
                        val jsonString = client.discoverServices(
                            deviceIdentifier = call.argument<String>(ArgumentKey.DEVICE_IDENTIFIER)!!,
                            serviceUUIDs = serviceUuids
                        )
                        result.success(jsonString)
                    }
                    MethodName.DISCOVER_CHARACTERISTICS -> {
                        val characteristicsUuids =
                            call.argument<List<String>>(ArgumentKey.CHARACTERISTIC_UUIDS)?.map {
                                UUIDfrom(bluetoothUUIDStr = it)
                            }
                        val jsonString = client.discoverCharacteristics(
                            deviceIdentifier = call.argument<String>(ArgumentKey.DEVICE_IDENTIFIER)!!,
                            serviceUuid = UUIDfrom(call.argument<String>(ArgumentKey.SERVICE_UUID)!!),
                            characteristicsUuids = characteristicsUuids
                        )
                        result.success(jsonString)
                    }
                    MethodName.GET_CHARACTERISTICS -> {
                        val jsonString = client.characteristics(
                            deviceIdentifier = call.argument<String>(ArgumentKey.DEVICE_IDENTIFIER)!!,
                            serviceUuid = UUIDfrom(call.argument<String>(ArgumentKey.SERVICE_UUID)!!)
                        )
                        result.success(jsonString)
                    }
                    MethodName.GET_SERVICES -> {
                        val jsonString = client.services(
                            deviceIdentifier = call.argument<String>(ArgumentKey.DEVICE_IDENTIFIER)!!
                        )
                        result.success(jsonString)
                    }
                    MethodName.GET_DESCRIPTORS_FOR_DEVICE -> {
                        val jsonString = client.descriptorsForDevice(
                            deviceIdentifier = call.argument<String>(ArgumentKey.DEVICE_IDENTIFIER)!!,
                            serviceUuid = UUIDfrom(call.argument<String>(ArgumentKey.SERVICE_UUID)!!),
                            characteristicUuid = UUIDfrom(call.argument<String>(ArgumentKey.CHARACTERISTIC_UUID)!!)
                        )
                        result.success(jsonString)
                    }
                    MethodName.RSSI -> {
                        val rssi = client.rssi(
                            deviceIdentifier = call.argument<String>(ArgumentKey.DEVICE_IDENTIFIER)!!
                        )
                        result.success(rssi)
                    }
                    MethodName.REQUEST_MTU -> {
                        val newMTU = client.requestMtu(
                            deviceIdentifier = call.argument<String>(ArgumentKey.DEVICE_IDENTIFIER)!!,
                            mtu = call.argument<Int>(ArgumentKey.MTU)!!
                        )
                        result.success(newMTU)
                    }
                    MethodName.GET_CONNECTED_DEVICES -> {
                        val serviceUUIDs = call.argument<List<String>>(ArgumentKey.UUIDS)!!.map {
                            UUIDfrom(bluetoothUUIDStr = it)
                        }
                        val jsonString = client.getConnectedDevices(
                            serviceUUIDs = serviceUUIDs
                        )
                        result.success(jsonString)
                    }
                    MethodName.GET_KNOWN_DEVICES -> {
                        val jsonString = client.getKnownDevices(
                            deviceIdentifiers = call.argument<List<String>>(ArgumentKey.DEVICE_IDENTIFIERS)!!
                        )
                        result.success(jsonString)
                    }
                    MethodName.READ_CHARACTERISTIC_FOR_DEVICE -> {
                        val jsonString = client.readCharacteristicForDevice(
                            deviceIdentifier = call.argument<String>(ArgumentKey.DEVICE_IDENTIFIER)!!,
                            serviceUuidStr = call.argument<String>(ArgumentKey.SERVICE_UUID)!!,
                            characteristicUuidStr = call.argument<String>(ArgumentKey.CHARACTERISTIC_UUID)!!,
                        )
                        result.success(jsonString)
                    }
                    MethodName.WRITE_CHARACTERISTIC_FOR_DEVICE -> {
                        val withResponse =
                            call.argument<Boolean>(ArgumentKey.WITH_RESPONSE) ?: false;
                        val jsonString =  client.writeCharacteristicForDevice(
                            deviceIdentifier = call.argument<String>(ArgumentKey.DEVICE_IDENTIFIER)!!,
                            serviceUuidStr = call.argument<String>(ArgumentKey.SERVICE_UUID)!!,
                            characteristicUuidStr = call.argument<String>(ArgumentKey.CHARACTERISTIC_UUID)!!,
                            bytesToWrite = call.argument<ByteArray>(ArgumentKey.VALUE)!!,
                            withResponse = withResponse
                        )
                        result.success(jsonString)
                    }
                    MethodName.MONITOR_CHARACTERISTIC_FOR_DEVICE -> {
                        val key = client.monitorCharacteristicForDevice(
                            deviceIdentifier = call.argument<String>(ArgumentKey.DEVICE_IDENTIFIER)!!,
                            serviceUuid = UUIDfrom(call.argument<String>(ArgumentKey.SERVICE_UUID)!!),
                            characteristicUuid = UUIDfrom(call.argument<String>(ArgumentKey.CHARACTERISTIC_UUID)!!)
                        )
                        result.success(key)
                    }
                    MethodName.READ_DESCRIPTOR_FOR_DEVICE -> {
                        val jsonString = client.readDescriptorForDevice(
                            deviceIdentifier = call.argument<String>(ArgumentKey.DEVICE_IDENTIFIER)!!,
                            serviceUuidStr = call.argument<String>(ArgumentKey.SERVICE_UUID)!!,
                            characteristicUuidStr = call.argument<String>(ArgumentKey.CHARACTERISTIC_UUID)!!,
                            descriptorUuidStr = call.argument<String>(ArgumentKey.DESCRIPTOR_UUID)!!
                        )
                        result.success(jsonString)
                    }
                    MethodName.WRITE_DESCRIPTOR_FOR_DEVICE -> {
                        val jsonString =  client.writeDescriptorForDevice(
                            deviceIdentifier = call.argument<String>(ArgumentKey.DEVICE_IDENTIFIER)!!,
                            serviceUuidStr = call.argument<String>(ArgumentKey.SERVICE_UUID)!!,
                            characteristicUuidStr = call.argument<String>(ArgumentKey.CHARACTERISTIC_UUID)!!,
                            descriptorUuidStr = call.argument(ArgumentKey.DESCRIPTOR_UUID)!!,
                            value = call.argument<ByteArray>(ArgumentKey.VALUE)!!
                        )
                        result.success(jsonString)
                    }
                    else -> throw NotImplementedError()
                }
            } catch (e: NotImplementedError) {
                result.notImplemented()
            } catch (e: Throwable) {
                result.error(throwable = e)
            }
        }
    }
}

fun MethodChannel.Result.error(throwable: Throwable) {
    if (throwable is BleError) {
        error(bleError = throwable)
        return
    }
    error(BleErrorCode.UnknownError.code.toString(), throwable.localizedMessage, throwable.cause)
}

fun MethodChannel.Result.error(bleError: BleError) {
    error(bleError.errorCode.code.toString(), bleError.message, bleError.cause)
}